src/data/ProgramsD2Repository.ts

```diff
@@ -0,0 +1,185 @@
+                            : []
+                    );
+                    const currentDate = new Date();
+                    currentDate.setDate(currentDate.getDate() - daysOfReference);
+                    const occurredBefore = currentDate.getTime();
+                    return (
+                        dates &&
+                        !_.isEmpty(dates) &&
+                        Math.max(Math.max(...dates), occurredBefore) === occurredBefore
+                    );
+                });
+                return filteredEntities;
+            });
+
+        const payload$ = filteredEntities$.map(({ data: teis }) => {
+            const enrollmentsWithLastDate = teis.flatMap(tei => {
+                const e = _.first(tei.enrollments);
+                if (!e) return [];
+                const { orgUnit, program, trackedEntity, enrollment, enrolledAt, occurredAt } = e;
+                const dates = e.events.flatMap(event =>
+                    programStagesIds.includes(event.programStage)
+                        ? [new Date(event.occurredAt).getTime()]
+                        : []
+                );
+                return [
+                    {
+                        enrollment: {
+                            orgUnit,
+                            program,
+                            trackedEntity,
+                            enrollment,
+                            enrolledAt,
+                            occurredAt,
+                            status: "COMPLETED",
+                        },
+                        lastConsultationDate: dates && !_.isEmpty(dates) && Math.max(...dates),
+                    },
+                ];
+            });
+
+            const events = enrollmentsWithLastDate.flatMap(
+                ({ enrollment: e, lastConsultationDate: date }) => {
+                    if (!date) return [];
+                    const ocurredAt = new Date(date);
+                    ocurredAt.setDate(ocurredAt.getDate() + daysOfReference);
+                    const dataValues = pairsDeValue.map(([dataElement, value]) => ({ dataElement, value }));
+                    const [commentDe, commentValue] = comments ?? [];
+                    return [
+                        {
+                            status: e.status,
+                            programStage: closureProgramId,
+                            enrollment: e.enrollment,
+                            orgUnit: e.orgUnit,
+                            occurredAt: ocurredAt.toISOString(),
+                            createdAt: new Date().toISOString(),
+                            updatedAt: new Date().toISOString(),
+                            dataValues:
+                                commentDe && commentValue
+                                    ? [...dataValues, { dataElement: commentDe, value: commentValue }]
+                                    : dataValues,
+                        },
+                    ];
+                }
+            );
```

Comment:
The body of this method is quite big. The usual approach: abstract code into pieces (private class methods are usually ok). This both reduces the line count of the body and gives names to the actions. Ideally, the main public method should fit in a screen and be so declarative that a non-programmer understands what it's doing.